### PR TITLE
Improved param and rx docs

### PR DESCRIPTION
**Pull request by MarcSkovMadsen**
Attempt to make `.param`, `.rx` and `.rx()` easier to use by improving docstrings.

Currently I meet a lot of dead ends because of missing or very sparse docstrings.

For example `.rx` in example below has no docstring which makes it almost impossible to use without already knowing how.

```python
import param

class MyClass(param.Parameterized):
    value = param.Parameter()

my_class = MyClass(value=0)
rx_value = my_class.param.value.rx
```

Its the same for `.param`. There is no docstring.

I do believe it would help users tremendously if there where docstring and improved docstrings. It would make the `.param` and `.rx` namespaces so much more inviting and easy to use.

## Guiding Principles

- Use Numpy style docstrings (would strongly prefer google docstrings because they are better as tooltips)
- Add types where they are easy to add
- Document in style similar to Panel docstrings
  - Reference link
  - Basic examples that can run
- Ok to make smaller refactors of code if they can be done without breaking changes.

### CODE REVIEW

File: param/reactive.py
Comment (MarcSkovMadsen) [inline param/reactive.py:790]: This enables me to provide a docstring. Don't know how to do that without it.

**Comment by MarcSkovMadsen** [comment]
As I am not a regular contributor to param it would help me to get a review, feedback and merge of this "smaller" PR.

Then I would feel confident making a larger PR more systematically going through `.param` and `.rx` namespaces and improving the docstrings and types.

File: param/parameterized.py
Comment (MarcSkovMadsen) [inline param/parameterized.py:2758]: Now I better understand why I've never felt confident starting to use some of these methods.

They feel so "unsupported" when there is no useful docstring. And without types you have to experiment to figure out what goes in and comes out.

File: .gitignore
Comment (hoxbro) [inline .gitignore:39]:
```suggestion
```

I don't think personal files like this should be added to the `.gitignore`.

You can add it to a global `.gitignore` on your computer, this is what I do, an example of how to set it up [here](https://sebastiandedeyne.com/setting-up-a-global-gitignore-file/).

**Review by hoxbro — COMMENTED** [review]
Thank you so much! Love it.

File: param/parameterized.py
Comment (MarcSkovMadsen) [reply param/parameterized.py:2297]: I don't think so. The tests pass?

File: param/reactive.py
Comment (MarcSkovMadsen) [reply param/reactive.py:787]: `reactive_ops` (reactive operators) and `rx` (reactive expressions) are two different classes and should not have the same docstring. I actually did not understand there was a difference before I started to inspect the code and add docstrings.

Comment (MarcSkovMadsen) [reply param/reactive.py:788]: I added pytests for the 3 docstrings that should be identical. Thx.

File: param/parameterized.py
Comment (hoxbro) [reply param/parameterized.py:2298]: More to postpone the evaluation of the annotations, https://peps.python.org/pep-0563/. 

My original thinking was it was faster with that enabled, but I haven't tested it.

**Review by hoxbro — APPROVED** [review]
Let's see if @philippjfr has anything to say.

File: .gitignore
Comment (MarcSkovMadsen) [reply .gitignore]: removed

File: tests/testreactive.py
Comment (hoxbro) [inline tests/testreactive.py:772]:
```suggestion
def test_docstrings_in_sync():
     # The docstring needs to be explicitly written to work with LSP. 
```

File: param/parameterized.py
Comment (maximlt) [inline param/parameterized.py:2327]: "to the class via the class"?

File: param/parameterized.py
Comment (maximlt) [inline param/parameterized.py:2394]: Param has an API reference. What do you think about linking to the ref instead of to the user guide?

https://param.holoviz.org/reference/param/generated/param.parameterized.Parameters.update.html

File: param/parameterized.py
Comment (maximlt) [inline param/parameterized.py:2707]: How about just typing the return value of `JSONSerialization.serialize_parameters`? At the moment it seems only JSON is supported here so `str` is correct at this level but other formats may be added in the future.

File: param/parameterized.py
Comment (maximlt) [inline param/parameterized.py:2722]: That's not a dict.

File: param/parameterized.py
Comment (maximlt) [inline param/parameterized.py:2396]: Shouldn't this be formatted as follows?

```suggestion
        Examples
        --------
```

File: param/parameterized.py
Comment (maximlt) [inline param/parameterized.py:1331]:
```suggestion
        Examples
        --------
```

File: param/parameterized.py
Comment (maximlt) [inline param/parameterized.py:2313]:
```suggestion
        Examples
        --------
```

File: param/parameterized.py
Comment (maximlt) [inline param/parameterized.py:2712]: Seems to be that's not how Numpy Docstrings are usually formatted? Based on https://numpydoc.readthedocs.io/en/latest/format.html#parameters.

```suggestion
        subset: list, optional
```

File: param/parameterized.py
Comment (maximlt) [inline param/parameterized.py:2718]: Personal opinion but I usually don't find listing the errors raised super useful.

File: param/parameterized.py
Comment (maximlt) [inline param/parameterized.py:2729]:
```suggestion
        Examples
        --------
```

File: param/parameterized.py
Comment (maximlt) [inline param/parameterized.py:2764]: This docstring is not formatted following the Numpy style.

**Review by maximlt — CHANGES_REQUESTED** [review]
I found inconsistencies in the formatting of the docstrings added in this PR. I also wonder if the links shouldn't point to the reference API instead of to the user guides.

Before merging, I'd recommend building the dev site from this branch and checking that the generated API ref is correctly formatted.

File: param/parameterized.py
Comment (philippjfr) [reply param/parameterized.py:2394]: I probably wouldn't call this `Reference:`, rather I'd suggest `Read more:` or something. The user guide is going to be more helpful than the reference guide here though since the API docs will just be generated from this docstring.

**Review by philippjfr — CHANGES_REQUESTED** [review]
This is well overdue, so thanks for doing it. Let's make sure that we consistently use NumPy style docstrings though, right now this is an odd mix.

**Comment by MarcSkovMadsen** [comment]
> This is well overdue, so thanks for doing it. Let's make sure that we consistently use NumPy style docstrings though, right now this is an odd mix.

If think I should investigate whether ruff or something else can validate this. I'm not using Numpy Docstrings very much. So I will fail at this.

**Comment by MarcSkovMadsen** [comment]
Thanks so much for the feedback. Will correct it ASAP.

File: param/parameterized.py
Comment (MarcSkovMadsen) [reply param/parameterized.py:2327]: How would you formulate it.

File: param/parameterized.py
Comment (MarcSkovMadsen) [reply param/parameterized.py:2708]: I've removed the type annotation.

**Comment by MarcSkovMadsen** [comment]
> This is well overdue, so thanks for doing it. Let's make sure that we consistently use NumPy style docstrings though, right now this is an odd mix.

I've updated the docstrings. Also means the code blocks are now `>>>` prefixed.

File: param/parameterized.py
Comment (maximlt) [inline param/parameterized.py:4411]:
```suggestion
        >>> my_instance.value  # the current parameter value
        0
        """
```

File: param/parameterized.py
Comment (maximlt) [inline param/parameterized.py:1331]: Is this required?

File: param/parameterized.py
Comment (maximlt) [inline param/parameterized.py:1350]: Looks like `b` is not needed for this example

```suggestion
        >>> p = P(a=1)
```

File: param/parameterized.py
Comment (maximlt) [inline param/parameterized.py:2392]:
```suggestion
        User Guide
        ----------
        https://param.holoviz.org/user_guide/Parameters.html#other-parameterized-methods
```

File: param/parameterized.py
Comment (maximlt) [inline param/parameterized.py:1349]:
```suggestion
        ...     a = param.Number()
        ...     b = param.String()
```

File: param/parameterized.py
Comment (maximlt) [inline param/parameterized.py:2321]:
```suggestion
        ...     a = param.Number()
        ...     b = param.String()
```

File: param/parameterized.py
Comment (maximlt) [inline param/parameterized.py:2401]:
```suggestion
        ...     a = param.Number()
        ...     b = param.String()
```

File: param/parameterized.py
Comment (maximlt) [inline param/parameterized.py:2727]:
```suggestion
        ...     a = param.Number()
        ...     b = param.String()
```

File: param/parameterized.py
Comment (maximlt) [inline param/parameterized.py:2327]:
```suggestion
        >>> p.c
```

File: param/parameterized.py
Comment (maximlt) [inline param/parameterized.py:2333]:
```suggestion
        >>> p.d
```

File: param/parameterized.py
Comment (maximlt) [reply param/parameterized.py:2324]: Ok I see what you mean now, I'll make suggestions to make it just a little bit more explicit.

File: param/parameterized.py
Comment (maximlt) [inline param/parameterized.py:2324]:
```suggestion
        Add a new parameter to the class `P` via the class namespace `P.param`:
```

File: param/parameterized.py
Comment (maximlt) [inline param/parameterized.py:2330]:
```suggestion
        Add a new parameter to the class `P` via the instance namespace `p.param`:
```

**Review by maximlt — CHANGES_REQUESTED** [review]
Thanks for the updates Marc!

I see that Panel also has its examples code doctest formatted (`>>> <code>`). What's your experience using this format? I know that it sometimes annoys me as copy/pasting it (from a website, a docstring tooltip in an IDE, etc.) doesn't always strip `>>>` so you have to manually remove them. IIRC the `ipython` interpreter strips them out but not the base python REPL. The advantage of this format is that it enables running doctest which might be a good idea. By the way, I have no strong opinion over all of that, just trying to gauge the user experience.

**Comment by MarcSkovMadsen** [comment]
I also Think the are hard to copy. But I changed them because of request to Alice with nu not docstrings. I believe that is the format.

**Comment by MarcSkovMadsen** [comment]
It is the format

image

**Comment by MarcSkovMadsen** [comment]
I believe I got the feedback from @maximlt above that `>>>` codestyle is not very helpful as its hard to copy examples. I agree.

So even though `>>>` is numpy docstyle. I plan to overrule this by restructuredText code blocks. Please let me know if you disagree @hoxbro , @philippjfr,  @maximlt as implementing one style and then have to change it is a massive task. Thanks.

**Comment by maximlt** [comment]
> So even though >>> is numpy docstyle. I plan to overrule this by restructuredText code blocks.

Could you show what that format looks like in the code? I'd also like to know how it will end up being displayed:

- in VSCode
- in JupyterLab
- on the reference API of Param's website (you could just make a change to your branch, trigger a dev doc build from Github, and post a screenshot)

**Comment by MarcSkovMadsen** [comment]
> > So even though >>> is numpy docstyle. I plan to overrule this by restructuredText code blocks.
> 
> Could you show what that format looks like in the code? I'd also like to know how it will end up being displayed:
> 
> * in VSCode
> * in JupyterLab
> * on the reference API of Param's website (you could just make a change to your branch, trigger a dev doc build from Github, and post a screenshot)

I'm simply not prepared to do that work again. I tried to analyze situation when started doing this for Panel. And the learning is there is no way that works great. I can focus on VS Code to be honest. The other part is too much.

**Comment by maximlt** [comment]
> I'm simply not prepared to do that work again. I tried to analyze situation when started doing this for Panel. And the learning is there is no way that works great. If you are happy with Panel I'll do it like that. That is what I can offer.

Ok sure I understand, this is all a bit of a mess.

> So even though `>>>` is numpy docstyle. I plan to overrule this by restructuredText code blocks.

Could you show how you plan to change it with an example? For example, how will this docstring look like with the change you plan to make?

```
        """
        Return the serialized parameters of the Parameterized object.

        Parameters
        ----------
        subset : list, optional
            A list of parameter names to serialize. If None, all parameters will be serialized. Defaults to None.
        mode : str, optional
            The serialization format. By default, only 'json' is supported. Defaults to 'json'.

        Returns
        -------
        Any
            The serialized value.

        User Guide
        ----------
        https://param.holoviz.org/user_guide/Serialization_and_Persistence.html#serializing-with-json

        Examples
        --------
        Create a Parameterized instance and serialize its parameters:

        >>> import param
        >>> class P(param.Parameterized):
        >>>     a = param.Number()
        >>>     b = param.String()
        >>> p = P(a=1, b="hello")

        Serialize parameters:

        >>> serialized_data = p.param.serialize_parameters()
        >>> print(serialized_data)
        {"name": "P00002", "a": 1, "b": "hello"}
        """
```

Then I can look a little bit what would the user experience be on various platforms (VSCode, JLab, website, etc.).

---

Just linking something that I wasn't aware of, the `sphinx-copybutton` Sphinx extension has lots of ways to configure what gets copied in the end, possibly stripping out `>>> `: https://sphinx-copybutton.readthedocs.io/en/latest/use.html. I'm not sure why Pandas doesn't use that (check https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.sort_values.html).

File: param/parameterized.py
Comment (jbednar) [inline param/parameterized.py:2383]: Does `update` let you set a bunch of parameter values at once without triggering events until all of them have been set? I may be confused about some other method, but if this is what lets you do that, it should presumably be mentioned in the docstring.

**Review by jbednar — APPROVED** [review]
This looks very useful; thanks @MarcSkovMadsen !

File: param/parameterized.py
Comment (maximlt) [reply param/parameterized.py:2383]: Yes, usually we recommend updating multiple parameters at once with `update` instead of sequentially, specially in a Panel app where this can lead to multiple re-renders.

Comment (jbednar) [reply param/parameterized.py:2383]: I'm not sure how best to express that in terms that make sense at the Param level, and didn't see much in the code that would let me state precisely what is deferred during the update.  @philippjfr, can we say that events are not delivered to watchers until all parameter values are updated?

File: param/parameterized.py
Comment (MarcSkovMadsen) [reply param/parameterized.py:1350]: I've improved it even further in my code. Thx for the suggestion.

File: param/parameterized.py
Comment (MarcSkovMadsen) [reply param/parameterized.py:1349]: fixed in my version.

File: param/parameterized.py
Comment (MarcSkovMadsen) [reply param/parameterized.py:2401]: fixed in my version. Thx.

File: param/parameterized.py
Comment (MarcSkovMadsen) [reply param/parameterized.py:2727]: Fixed in my version. Thx.

File: param/parameterized.py
Comment (MarcSkovMadsen) [reply param/parameterized.py:2333]: Fixed in my version. Thx.

File: param/parameterized.py
Comment (MarcSkovMadsen) [reply param/parameterized.py:2324]: added in my version. Thx a lot.

File: param/parameterized.py
Comment (MarcSkovMadsen) [reply param/parameterized.py:2330]: Added in my version. Great suggestion. Thx.

File: param/parameterized.py
Comment (MarcSkovMadsen) [reply param/parameterized.py:2383]: > Yes, usually we recommend updating multiple parameters at once with `update` instead of sequentially, specially in a Panel app where this can lead to multiple re-renders.

I'm always in doubt about `.update`. Does it batch updates. I've never seen that explained or demonstrated.

**Comment by MarcSkovMadsen** [comment]
> > I'm simply not prepared to do that work again. I tried to analyze situation when started doing this for Panel. And the learning is there is no way that works great. If you are happy with Panel I'll do it like that. That is what I can offer.
> 
> Ok sure I understand, this is all a bit of a mess.
> 
> > So even though `>>>` is numpy docstyle. I plan to overrule this by restructuredText code blocks.
> 
> Could you show how you plan to change it with an example? For example, how will this docstring look like with the change you plan to make?
> 
> ```
>         """
>         Return the serialized parameters of the Parameterized object.
> 
>         Parameters
>         ----------
>         subset : list, optional
>             A list of parameter names to serialize. If None, all parameters will be serialized. Defaults to None.
>         mode : str, optional
>             The serialization format. By default, only 'json' is supported. Defaults to 'json'.
> 
>         Returns
>         -------
>         Any
>             The serialized value.
> 
>         User Guide
>         ----------
>         https://param.holoviz.org/user_guide/Serialization_and_Persistence.html#serializing-with-json
> 
>         Examples
>         --------
>         Create a Parameterized instance and serialize its parameters:
> 
>         >>> import param
>         >>> class P(param.Parameterized):
>         >>>     a = param.Number()
>         >>>     b = param.String()
>         >>> p = P(a=1, b="hello")
> 
>         Serialize parameters:
> 
>         >>> serialized_data = p.param.serialize_parameters()
>         >>> print(serialized_data)
>         {"name": "P00002", "a": 1, "b": "hello"}
>         """
> ```
> 
> Then I can look a little bit what would the user experience be on various platforms (VSCode, JLab, website, etc.).
> 
> Just linking something that I wasn't aware of, the `sphinx-copybutton` Sphinx extension has lots of ways to configure what gets copied in the end, possibly stripping out `>>> `: https://sphinx-copybutton.readthedocs.io/en/latest/use.html. I'm not sure why Pandas doesn't use that (check https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.sort_values.html).

The below is how it looks in my version of VS Code on windows with pylance. I don't have the latest version of VS Code.

https://github.com/user-attachments/assets/170a9117-ccbc-43ad-8ed9-950b4fac899e

File: param/parameterized.py
Comment (MarcSkovMadsen) [reply param/parameterized.py:1331]: I've removed it.

File: param/parameterized.py
Comment (MarcSkovMadsen) [reply param/parameterized.py:2382]: I does update everything and then execute callbacks:

```python
import param

class MyParam(param.Parameterized):
    a = param.String()
    b = param.String()

my_param = MyParam(a="0. Hello", b="0. World")

@param.depends(my_param.param.a, watch=True)
def print_b(a):
    print(my_param.b)

my_param.a="1. Hello"
my_param.b="1. World"
my_param.param.update(a="2. Hello",b="2. World")
```

```bash
0. World
2. World
```

Comment (MarcSkovMadsen) [reply param/parameterized.py:2378]: I*ve updated the docstring based on the feedback.

**Review by maximlt — CHANGES_REQUESTED** [review]
LGTM other than 1 pending suggestion. Let's preserve the original Numpy style for the examples, I think we can tweak the sphinx-copybutton extension to work in a decent way, and hopefully VSCode will make it easier to copy/paste these snippets. If we end up needing to change the style, it shouldn't be too hard to write a script to re-write the docstrings automatically, stripping `>>>` and `...` (the other way around, yes!).

**Pull request merged**
